components/src/jsMain/kotlin/dev/fritz2/components/validation/validation.kt

```diff
@@ -0,0 +1,38 @@
+package dev.fritz2.components.validation
+
+import dev.fritz2.binding.RootStore
+import dev.fritz2.binding.Store
+import dev.fritz2.binding.SubStore
+import dev.fritz2.validation.Validator
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.map
+
+interface Validatable<D, T> : Store<D> {
+    val validator: Validator<D, ComponentValidationMessage, T>
+    fun validate(metadata: T) {
+        syncBy(handle<D> { old, new ->
+            if (validator.isValid(new, metadata)) new else old
+        })
+    }
+}
+
+class RootStoreIsNotValidatable(id: String) :
+    Exception("RootStore of data with id=$id must implement Validatable interface")
+
+fun <D> Store<D>.validationMessage(): Flow<String> = when (this) {
+    is RootStore<*> -> {
```

Comment:
You could declare this method here abstract and implement it at RootStore and SubStore. This would fit better to all the other methods that have different behavior for Root and Sub.

Reply:
I don't want to change our core implementation for a special case in components. In my opinion this extension method approach fits the needs and provides a better separation of the different code bases.